### PR TITLE
refactor(Tree): remove context from TreeItem

### PR DESCRIPTION
**Pull request by GoranRibic**
Remove as much as possible usage of context in TreeItem, memoize components, and reduce the number of re-renders.

### CODE REVIEW

File: src/components/Tree/helpers/nodes.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+export const removeNodesFromTree = (tree: ReactElement[], nodeIds: string[]): ReactElement[] => {
+    // Create a set of the node IDs to remove for faster lookup
+    const nodesToRemove = new Set(nodeIds);
+
+    // Filter the tree array to remove the nodes with IDs in the set
+    return tree.filter((node) => !nodesToRemove.has(node.props.id));
+};
+
+export const getNodeIdsInBranch = (tree: ReactElement[], id: string): string[] => {
```
Comment (SamuelAlev) [inline src/components/Tree/helpers/nodes.ts:13]:
```suggestion
export const getNodeIdsInBranch = (tree: ReactElement[], startingNodeId: string): string[] => {
```

File: src/components/Tree/TreeItem/TreeItemOverlay.tsx [hunk 2]
```diff
@@ -1,27 +1,22 @@
...
+export type Overlay = {
+    id: string;
+    label?: string;
+    level?: number;
+    children: ReactNode;
+    contentComponent?: ReactNode;
+    isSelected?: boolean;
+};
 
+export const TreeItemOverlay = ({ id, label, level = 0, children, contentComponent, isSelected }: Overlay) => {
```
[the remark this answers is not in the document]
Comment (GoranRibic) [reply src/components/Tree/TreeItem/TreeItemOverlay.tsx:19]: It is needed for indentation calculation.

File: src/components/Tree/Tree.tsx [hunk 3]
```diff
@@ -47,504 +48,547 @@ const measuring: MeasuringConfiguration = {
...
         default:
-            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-            //@ts-ignore
             console.warn(`Updated tree with action "${action.type}" but it has not effect.`);
-            return state;
     }
-};
+});
+
+export const Tree = memo(
```
[the remark this answers is not in the document]
Comment (GoranRibic) [reply src/components/Tree/Tree.tsx:161]: Needed.

File: src/components/Tree/TreeItem/TreeItem.tsx [hunk 4]
```diff
@@ -74,7 +74,8 @@ export const TreeItem = memo(
 
         const isActive = active?.id === id;
 
-        const isWithin = projection !== null && over !== null && projection.depth > over.data.current?.level;
+        // eslint-disable-next-line eqeqeq
+        const isWithin = projection != null && over !== null && projection.depth > over.data.current?.level;
```
Comment (SamuelAlev) [inline src/components/Tree/TreeItem/TreeItem.tsx:78]: Not allowed

File: src/components/Tree/TreeItem/TreeItem.tsx [hunk 5]
```diff
@@ -34,12 +34,12 @@ type TreeItemPrivateProps = {
...
-    projection: Nullable<Projection>;
-    registerOverlay: (overlay: Overlay) => void;
-    unregisterNodeChildren: (payload: string) => void;
-    registerNodeChildren: (payload: RegisterNodeChildrenPayload) => void;
+    onSelect?: (id: string) => void;
+    onExpand?: (id: string) => void;
+    projection?: Nullable<Projection>;
+    registerOverlay?: (overlay: Overlay) => void;
+    unregisterNodeChildren?: (payload: string) => void;
+    registerNodeChildren?: (payload: RegisterNodeChildrenPayload) => void;
```
Comment (SamuelAlev) [inline src/components/Tree/TreeItem/TreeItem.tsx:42]: Why is it optional now?

File: src/components/Tree/TreeItem/TreeItemOverlay.tsx (hunk 2 above)
Comment (anxobotana) [inline src/components/Tree/TreeItem/TreeItemOverlay.tsx:19]: regarding `level`: `aria-level >= 1` like in _treeItem_ or remove it if it's not needed 

```
aria-level={level + 1}
```

**Review by SamuelAlev — APPROVED** [review]

**Pull request merged**